src/schema_registry/middlewares/prometheus.py

```diff
@@ -0,0 +1,53 @@
+"""
+Copyright (c) 2024 Aiven Ltd
+See LICENSE for details
+"""
+
+from collections.abc import Awaitable, Callable
+from dependency_injector.wiring import inject, Provide
+from fastapi import Depends, FastAPI, Request, Response
+from karapace.instrumentation.prometheus import PrometheusInstrumentation
+from schema_registry.container import SchemaRegistryContainer
+
+import logging
+import time
+
+LOG = logging.getLogger(__name__)
+
+
+@inject
+async def prometheus_middleware(
+    request: Request,
+    call_next: Callable[[Request], Awaitable[Response]],
+    prometheus: PrometheusInstrumentation = Depends(Provide[SchemaRegistryContainer.karapace_container.prometheus]),
+) -> Response:
+    # Set start time for request
+    setattr(request.state, prometheus.START_TIME_REQUEST_KEY, time.time())
+
+    # Extract request labels
+    path = request.url.path
```

Comment:
This can cause quite high cardinality, can we align this with the path template. E.g. `/subjects/<subject>/versions`?

Reply:
Yes I can think around it, it was similar thought I had on the tracing PR #1009, where I extract the resource from the path as I didn't find a way to get the template.

In prometheus tho, I do not think this should be a problem, major cardinality issues arise with too many labels. To instrument, might be necessary to see exactly the path that's being checked, but I also get your point. Let me reason around it and get back.

Reply:
Looking at this discussion - https://github.com/fastapi/fastapi/discussions/12975, the route template is only accessed after a successful response, I do need the path already to instrument the requests in progress. 
Unless we want to fallback and extract the base resource, i.e. `resource = request.url.path.split("/")[1]`. From a template like `/subjects/{subject}/versions` and a request for `/subjects/test-topic-value/versions`, this will yield `subjects`.